**<파일 경로>**
- C:/Users/Brian/Desktop/Data Science/Projects/Playdata/dataset/population/test

In [1]:
filepath = input()

C:/Users/Brian/Desktop/Data Science/Projects/Playdata/dataset/population/test


In [2]:
import numpy as np
import pandas as pd
from tqdm.notebook import tqdm

# 2020년 1~6월 인구 데이터 결합
<참고>
- 2020년도 인구 데이터 셋들은 2018~2019년도 데이터 셋들과 다르게, 빈 컬럼이 존재해서 데이터들을 불러올 때 오류가 발생했다.
- 때문에 Excel에서 빈 컬럼들을 하나하나 제거해 준 뒤에 데이터들을 불러왔다.

In [3]:
# 2020년 1~6월 인구 데이터
file_1 = pd.read_csv('{}/LOCAL_PEOPLE_DONG_202001.csv'.format(filepath))
file_2 = pd.read_csv('{}/LOCAL_PEOPLE_DONG_202002.csv'.format(filepath))
file_3 = pd.read_csv('{}/LOCAL_PEOPLE_DONG_202003.csv'.format(filepath))
file_4 = pd.read_csv('{}/LOCAL_PEOPLE_DONG_202004.csv'.format(filepath))
file_5 = pd.read_csv('{}/LOCAL_PEOPLE_DONG_202005.csv'.format(filepath))
file_6 = pd.read_csv('{}/LOCAL_PEOPLE_DONG_202006.csv'.format(filepath))

In [4]:
concat_list = [file_1, file_2, file_3, file_4, file_5, file_6]

pop_total = pd.concat(concat_list, axis = 0, ignore_index = True)
pop_total.head()

기준일ID  시간대구분     행정동코드      총생활인구수  남자0세부터9세생활인구수  남자10세부터14세생활인구수  \
0  20200101      0  11620625  24188.5806       870.1487         439.6648   
1  20200101      0  11140605   6716.5334        89.0565          36.2822   
2  20200101      0  11470640  46603.6243      2164.9685        1254.3274   
3  20200101      0  11290780   9589.1813       390.6835         244.4328   
4  20200101      0  11320512  29984.1309      1005.9128         506.6446   

   남자15세부터19세생활인구수  남자20세부터24세생활인구수  남자25세부터29세생활인구수  남자30세부터34세생활인구수  ...  \
0         453.6983         683.2883        1348.9122        1230.3016  ...   
1          78.8508         225.3331         416.2471         321.9111  ...   
2        1296.7883        1076.9660        1164.2301        1271.9820  ...   
3         192.6254         240.5164         288.9657         301.6838  ...   
4         637.8008         721.4613        1013.8781         972.6073  ...   

   여자25세부터29세생활인구수  여자30세부터34세생활인구수  여자35세부터39세생활인구수  여자40세부터44세생활인구수  \
0        1381.5367        1264.4003        1112.2324         889.4873   
1         439.5516         314.3750         209.5565         156.2364   
2        1318.5467        1510.0550        1877.1880        1947.7351   
3         273.5978         323.4285         395.5678         375.7008   
4        1079.7667        1071.6117        1332.6363        1053.3402   

   여자45세부터49세생활인구수  여자50세부터54세생활인구수  여자55세부터59세생활인구수  여자60세부터64세생활인구수  \
0         864.8068         832.7668         825.7571         777.3297   
1         171.9298         203.5514         248.0982         145.9047   
2        2149.1684        1894.8391        1964.5647        1812.9103   
3         416.7873         327.9834         380.6776         385.1141   
4        1345.4963        1144.1586        1397.3890        1362.8639   

   여자65세부터69세생활인구수  여자70세이상생활인구수  
0         708.1216     1501.7124  
1         139.5027      303.8293  
2        1473.4635     3226.1966  
3         311.8604      697.1448  
4         881.4148     2509.3527  

[5 rows x 32 columns]

In [5]:
pop_total.tail()

기준일ID  시간대구분     행정동코드      총생활인구수  남자0세부터9세생활인구수  \
1852027  20200630     23  11620615  17594.6090       324.0727   
1852028  20200630     23  11290575  23475.0694       506.7555   
1852029  20200630     23  11200520  17762.7968       616.1449   
1852030  20200630     23  11470680  28068.5528      1209.0730   
1852031  20200630     23  11350695  40382.4180      1651.3213   

         남자10세부터14세생활인구수  남자15세부터19세생활인구수  남자20세부터24세생활인구수  남자25세부터29세생활인구수  \
1852027         179.5686         289.7385         711.1608        1283.3871   
1852028         304.3084         588.4385        1209.3069        1381.8233   
1852029         299.9889         389.7595         414.0508         552.1344   
1852030         932.3942        1034.7436         725.5515         623.9077   
1852031         869.6293        1286.9991        1318.1163        1410.1635   

         남자30세부터34세생활인구수  ...  여자25세부터29세생활인구수  여자30세부터34세생활인구수  \
1852027        1187.2126  ...        1670.4222        1039.4546   
1852028         891.3540  ...        1731.8811        1081.4296   
1852029         619.2366  ...         773.5505         834.5563   
1852030         580.7503  ...         725.5334         719.6312   
1852031        1301.3931  ...        1530.0152        1476.5563   

         여자35세부터39세생활인구수  여자40세부터44세생활인구수  여자45세부터49세생활인구수  여자50세부터54세생활인구수  \
1852027         659.0542         504.4803         507.6602         462.9985   
1852028         830.1613         854.5790         894.3804         714.9320   
1852029        1008.0177         747.8779         743.3067         613.6456   
1852030        1187.0620        1329.0801        1426.4283        1145.3341   
1852031        1630.1772        1591.9283        1836.2814        1683.7437   

         여자55세부터59세생활인구수  여자60세부터64세생활인구수  여자65세부터69세생활인구수  여자70세이상생활인구수  
1852027         570.1290         468.1912         428.2289     1107.0345  
1852028         672.4728         530.3269         403.7214     1308.7104  
1852029         800.3507         636.8116         506.5551     1194.4720  
1852030        1146.4765         983.0643         785.6000     1932.5491  
1852031        1635.7655        1471.7540        1036.3136     3330.3663  

[5 rows x 32 columns]

# 기준일ID 변수 타입 변경
- int 타입으로 되어있는 "기준일ID" 변수를 먼저 str 타입으로 변경해 준 뒤, 다시 datetime 타입으로 변경해주겠다.

In [6]:
pop_total.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1852032 entries, 0 to 1852031
Data columns (total 32 columns):
 #   Column           Dtype  
---  ------           -----  
 0   기준일ID            int64  
 1   시간대구분            int64  
 2   행정동코드            int64  
 3   총생활인구수           float64
 4   남자0세부터9세생활인구수    float64
 5   남자10세부터14세생활인구수  float64
 6   남자15세부터19세생활인구수  float64
 7   남자20세부터24세생활인구수  float64
 8   남자25세부터29세생활인구수  float64
 9   남자30세부터34세생활인구수  float64
 10  남자35세부터39세생활인구수  float64
 11  남자40세부터44세생활인구수  float64
 12  남자45세부터49세생활인구수  float64
 13  남자50세부터54세생활인구수  float64
 14  남자55세부터59세생활인구수  float64
 15  남자60세부터64세생활인구수  float64
 16  남자65세부터69세생활인구수  float64
 17  남자70세이상생활인구수     float64
 18  여자0세부터9세생활인구수    float64
 19  여자10세부터14세생활인구수  float64
 20  여자15세부터19세생활인구수  float64
 21  여자20세부터24세생활인구수  float64
 22  여자25세부터29세생활인구수  float64
 23  여자30세부터34세생활인구수  float64
 24  여자35세부터39세생활인구수  float64
 25  여자40세부터44세생활인구수  float64
 26  여자45세부터49세생활인구수  float64
 27  여자50세부터54세생활

In [7]:
# int to string
pop_total['기준일ID'] = pop_total['기준일ID'].astype(str)
# string to datetime
pop_total['기준일ID'] = pd.to_datetime(pop_total['기준일ID'], format = '%Y-%m-%d')
pop_total['기준일ID']

0         2020-01-01
1         2020-01-01
2         2020-01-01
3         2020-01-01
4         2020-01-01
             ...    
1852027   2020-06-30
1852028   2020-06-30
1852029   2020-06-30
1852030   2020-06-30
1852031   2020-06-30
Name: 기준일ID, Length: 1852032, dtype: datetime64[ns]

# 변수명 변경
- 한글로 되어 있는 변수명을 모두 영어로 변경해주겠다.

In [8]:
new_colnames = {'기준일ID':'date', '시간대구분':'hour', '행정동코드':'dong_code', '총생활인구수':'total', 
                '남자0세부터9세생활인구수':'m_0_9', '남자10세부터14세생활인구수':'m_10_14', '남자15세부터19세생활인구수':'m_15_19',
                '남자20세부터24세생활인구수':'m_20_24', '남자25세부터29세생활인구수':'m_25_29', '남자30세부터34세생활인구수':'m_30_34',
                '남자35세부터39세생활인구수':'m_35_39', '남자40세부터44세생활인구수':'m_40_44', '남자45세부터49세생활인구수':'m_45_49',
                '남자50세부터54세생활인구수':'m_50_54', '남자55세부터59세생활인구수':'m_55_59', '남자60세부터64세생활인구수':'m_60_64',
                '남자65세부터69세생활인구수':'m_65_69', '남자70세이상생활인구수':'m_over_70', '여자0세부터9세생활인구수':'f_0_9',
                '여자10세부터14세생활인구수':'f_10_14', '여자15세부터19세생활인구수':'f_15_19', '여자20세부터24세생활인구수':'f_20_24',
                '여자25세부터29세생활인구수':'f_25_29', '여자30세부터34세생활인구수':'f_30_34', '여자35세부터39세생활인구수':'f_35_39',
                '여자40세부터44세생활인구수':'f_40_44', '여자45세부터49세생활인구수':'f_45_49', '여자50세부터54세생활인구수':'f_50_54',
                '여자55세부터59세생활인구수':'f_55_59', '여자60세부터64세생활인구수':'f_60_64', '여자65세부터69세생활인구수':'f_65_69',
                '여자70세이상생활인구수':'f_over_70'}

pop_total.rename(columns = new_colnames, inplace = True)
pop_total.head()

date  hour  dong_code       total      m_0_9    m_10_14    m_15_19  \
0 2020-01-01     0   11620625  24188.5806   870.1487   439.6648   453.6983   
1 2020-01-01     0   11140605   6716.5334    89.0565    36.2822    78.8508   
2 2020-01-01     0   11470640  46603.6243  2164.9685  1254.3274  1296.7883   
3 2020-01-01     0   11290780   9589.1813   390.6835   244.4328   192.6254   
4 2020-01-01     0   11320512  29984.1309  1005.9128   506.6446   637.8008   

     m_20_24    m_25_29    m_30_34  ...    f_25_29    f_30_34    f_35_39  \
0   683.2883  1348.9122  1230.3016  ...  1381.5367  1264.4003  1112.2324   
1   225.3331   416.2471   321.9111  ...   439.5516   314.3750   209.5565   
2  1076.9660  1164.2301  1271.9820  ...  1318.5467  1510.0550  1877.1880   
3   240.5164   288.9657   301.6838  ...   273.5978   323.4285   395.5678   
4   721.4613  1013.8781   972.6073  ...  1079.7667  1071.6117  1332.6363   

     f_40_44    f_45_49    f_50_54    f_55_59    f_60_64    f_65_69  f_over_70  
0   889.4873   864.8068   832.7668   825.7571   777.3297   708.1216  1501.7124  
1   156.2364   171.9298   203.5514   248.0982   145.9047   139.5027   303.8293  
2  1947.7351  2149.1684  1894.8391  1964.5647  1812.9103  1473.4635  3226.1966  
3   375.7008   416.7873   327.9834   380.6776   385.1141   311.8604   697.1448  
4  1053.3402  1345.4963  1144.1586  1397.3890  1362.8639   881.4148  2509.3527  

[5 rows x 32 columns]

# 연령대별 그룹 범주화
- 남/여 구분 없이 연령대 별로 합쳐주겠다.
- 다음과 같이 총 7개의 그룹으로 범주화한 연령대 변수를 생성하겠다.
    - pop_below_20
        - 20세 미만
    - pop_20
        - 20대
    - pop_30
        - 30대
    - pop_40
        - 40대
    - pop_50
        - 50대
    - pop_60
        - 60대
    - pop_over_70
        - 70세 이상

In [9]:
# 총 7개의 그룹으로 범주화한 연령대 변수 생성
pop_total['pop_below_20'] = pop_total['m_0_9'] + pop_total['m_10_14'] + pop_total['m_15_19'] + pop_total['f_0_9'] + pop_total['f_10_14'] + pop_total['f_15_19']
pop_total['pop_20'] = pop_total['m_20_24'] + pop_total['m_25_29'] + pop_total['f_20_24'] + pop_total['f_25_29']
pop_total['pop_30'] = pop_total['m_30_34'] + pop_total['m_35_39'] + pop_total['f_30_34'] + pop_total['f_35_39']
pop_total['pop_40'] = pop_total['m_40_44'] + pop_total['m_45_49'] + pop_total['f_40_44'] + pop_total['f_45_49']
pop_total['pop_50'] = pop_total['m_50_54'] + pop_total['m_55_59'] + pop_total['f_50_54'] + pop_total['f_55_59']
pop_total['pop_60'] = pop_total['m_60_64'] + pop_total['m_65_69'] + pop_total['f_60_64'] + pop_total['f_65_69']
pop_total['pop_over_70'] = pop_total['m_over_70'] + pop_total['f_over_70']

# 필요한 변수들만 추출

In [10]:
pop_total = pop_total[['date', 'hour', 'dong_code', 'total', 'pop_below_20', 'pop_20', 
                       'pop_30', 'pop_40', 'pop_50', 'pop_60', 'pop_over_70']]
pop_total.head()

date  hour  dong_code       total  pop_below_20     pop_20     pop_30  \
0 2020-01-01     0   11620625  24188.5806     3347.7743  4405.0736  4848.7403   
1 2020-01-01     0   11140605   6716.5334      374.5654  1463.2070  1230.5230   
2 2020-01-01     0   11470640  46603.6243     9299.7225  4807.8480  6271.4931   
3 2020-01-01     0   11290780   9589.1813     1469.7041  1110.9373  1443.6632   
4 2020-01-01     0   11320512  29984.1309     3946.9892  3731.0167  4775.9897   

      pop_40     pop_50     pop_60  pop_over_70  
0  3635.8273  3035.2176  2479.2207    2436.7258  
1  1087.2124  1174.9897   735.8849     650.1513  
2  8053.9891  7231.2414  5797.2636    5142.0645  
3  1583.3635  1432.4918  1265.3663    1283.6551  
4  4731.1070  4726.8181  4032.7099    4039.5011

# 행정동코드 변환
- 인구 데이터의 "행정동코드" 변수가 한글이 아닌 숫자로 적혀있다.
- 때문에 행정동코드와 행정동을 하나하나 매칭시켜서 변경해주도록 하겠다.
    
**<참고>**
- **'11620685':'관악구 신사동'**은 딕셔너리에서 제외시켰다.
    - 지하철 역 주소 데이터를 보면, '신사동'은 '강남구'에만 존재하기 때문이다.

In [11]:
dic = {'11110515':'종로구 청운효자동', '11110530':'종로구 사직동', '11110540':'종로구 삼청동', '11110550':'종로구 부암동', 
       '11110560':'종로구 평창동', '11110570':'종로구 무악동', '11110580':'종로구 교남동', '11110600':'종로구 가회동', 
       '11110615':'종로구 종로1.2.3.4가동', '11110630':'종로구 종로5.6가동', '11110640':'종로구 이화동', '11110650':'종로구 혜화동', 
       '11110670':'종로구 창신제1동', '11110680':'종로구 창신제2동', '11110690':'종로구 창신제3동', '11110700':'종로구 숭인제1동', 
       '11110710':'종로구 숭인제2동', 
       
       '11140520':'중구 소공동', '11140540':'중구 회현동', '11140550':'중구 명동', '11140570':'중구 필동', '11140580':'중구 장충동', 
       '11140590':'중구 광희동', '11140605':'중구 을지로동', '11140615':'중구 신당동', '11140625':'중구 다산동', '11140635':'중구 약수동', 
       '11140645':'중구 청구동', '11140650':'중구 신당제5동', '11140665':'중구 동화동', '11140670':'중구 황학동', '11140680':'중구 중림동', 
       
       '11170510':'용산구 후암동', '11170520':'용산구 용산2가동', '11170530':'용산구 남영동', '11170555':'용산구 청파동', 
       '11170560':'용산구 원효로제1동', '11170570':'용산구 원효로제2동', '11170580':'용산구 효창동', '11170590':'용산구 용문동', 
       '11170625':'용산구 한강로동', '11170630':'용산구 이촌제1동', '11170640':'용산구 이촌제2동', '11170650':'용산구 이태원제1동', 
       '11170660':'용산구 이태원제2동', '11170685':'용산구 한남동', '11170690':'용산구 서빙고동', '11170700':'용산구 보광동',
       
       '11200520':'성동구 왕십리제2동', '11200535':'성동구 왕십리도선동', '11200540':'성동구 마장동', '11200550':'성동구 사근동',
       '11200560':'성동구 행당제1동', '11200570':'성동구 행당제2동', '11200580':'성동구 응봉동', '11200590':'성동구 금호1가동',
       '11200615':'성동구 금호2.3가동', '11200620':'성동구 금호4가동', '11200645':'성동구 옥수동', '11200650':'성동구 성수1가제1동', 
       '11200660':'성동구 성수1가제2동', '11200670':'성동구 성수2가제1동', '11200690':'성동구 성수2가제3동', '11200720':'성동구 송정동', 
       '11200790':'성동구 용답동',
       
       '11215710':'광진구 화양동', '11215730':'광진구 군자동', '11215740':'광진구 중곡제1동', '11215750':'광진구 중곡제2동',
       '11215760':'광진구 중곡제3동', '11215770':'광진구 중곡제4동', '11215780':'광진구 능동', '11215810':'광진구 광장동',
       '11215820':'광진구 자양제1동', '11215830':'광진구 자양제2동', '11215840':'광진구 자양제3동', '11215847':'광진구 자양제4동',
       '11215850':'광진구 구의제1동', '11215860':'광진구 구의제2동', '11215870':'광진구 구의제3동',
       
       '11230536':'동대문구 용신동', '11230545':'동대문구 제기동', '11230560':'동대문구 전농제1동', '11230570':'동대문구 전농제2동',
       '11230600':'동대문구 답십리제1동', '11230610':'동대문구 답십리제2동', '11230650':'동대문구 장안제1동', '11230660':'동대문구 장안제2동',
       '11230705':'동대문구 청량리동', '11230710':'동대문구 회기동', '11230720':'동대문구 휘경제1동', '11230730':'동대문구 휘경제2동',
       '11230740':'동대문구 이문제1동', '11230750':'동대문구 이문제2동',
      
       '11260520':'중랑구 면목제2동', '11260540':'중랑구 면목제4동', '11260550':'중랑구 면목제5동', '11260565':'중랑구 면목본동',
       '11260570':'중랑구 면목제7동', '11260575':'중랑구 면목제3.8동', '11260580':'중랑구 상봉제1동', '11260590':'중랑구 상봉제2동', 
       '11260600':'중랑구 중화제1동', '11260610':'중랑구 중화제2동', '11260620':'중랑구 묵제1동', '11260630':'중랑구 묵제2동', 
       '11260655':'중랑구 망우본동', '11260660':'중랑구 망우제3동', '11260680':'중랑구 신내1동', '11260690':'중랑구 신내2동',
      
       '11290525':'성북구 성북동', '11290555':'성북구 삼선동', '11290575':'성북구 동선동', '11290580':'성북구 돈암제1동', 
       '11290590':'성북구 돈암제2동', '11290600':'성북구 안암동', '11290610':'성북구 보문동', '11290620':'성북구 정릉제1동', 
       '11290630':'성북구 정릉제2동', '11290640':'성북구 정릉제3동', '11290650':'성북구 정릉제4동', '11290660':'성북구 길음제1동', 
       '11290685':'성북구 길음제2동', '11290705':'성북구 종암동', '11290715':'성북구 월곡제1동', '11290725':'성북구 월곡제2동', 
       '11290760':'성북구 장위제1동', '11290770':'성북구 장위제2동', '11290780':'성북구 장위제3동', '11290810':'성북구 석관동',
       
       '11305534':'강북구 삼양동', '11305535':'강북구 미아동', '11305545':'강북구 송중동', '11305555':'강북구 송천동',
       '11305575':'강북구 삼각산동', '11305590':'강북구 번제1동', '11305600':'강북구 번제2동', '11305606':'강북구 번제3동',
       '11305610':'강북구 수유제1동', '11305620':'강북구 수유제2동', '11305630':'강북구 수유제3동', '11305645':'강북구 우이동', 
       '11305660':'강북구 인수동',
      
       '11320511':'도봉구 창제1동', '11320512':'도봉구 창제2동', '11320513':'도봉구 창제3동', '11320514':'도봉구 창제4동',
       '11320515':'도봉구 창제5동', '11320521':'도봉구 도봉제1동', '11320522':'도봉구 도봉제2동', '11320660':'도봉구 쌍문제1동',
       '11320670':'도봉구 쌍문제2동', '11320680':'도봉구 쌍문제3동', '11320681':'도봉구 쌍문제4동', '11320690':'도봉구 방학제1동',
       '11320700':'도봉구 방학제2동', '11320710':'도봉구 방학제3동',
      
       '11350560':'노원구 월계1동', '11350570':'노원구 월계2동', '11350580':'노원구 월계3동', '11350595':'노원구 공릉1동',
       '11350600':'노원구 공릉2동', '11350611':'노원구 하계1동', '11350612':'노원구 하계2동', '11350619':'노원구 중계본동',
       '11350621':'노원구 중계1동', '11350624':'노원구 중계4동', '11350625':'노원구 중계2.3동', '11350630':'노원구 상계1동', 
       '11350640':'노원구 상계2동', '11350665':'노원구 상계3.4동', '11350670':'노원구 상계5동', '11350695':'노원구 상계6.7동', 
       '11350700':'노원구 상계8동', '11350710':'노원구 상계9동', '11350720':'노원구 상계10동',
      
       '11380510':'은평구 녹번동', '11380520':'은평구 불광제1동', '11380530':'은평구 불광제2동', '11380551':'은평구 갈현제1동',
       '11380552':'은평구 갈현제2동', '11380560':'은평구 구산동', '11380570':'은평구 대조동', '11380580':'은평구 응암제1동',
       '11380590':'은평구 응암제2동', '11380600':'은평구 응암제3동', '11380625':'은평구 역촌동', '11380631':'은평구 신사제1동',
       '11380632':'은평구 신사제2동', '11380640':'은평구 증산동', '11380650':'은평구 수색동', '11380690':'은평구 진관동',
      
       '11410520':'서대문구 천연동', '11410555':'서대문구 북아현동', '11410565':'서대문구 충현동', '11410585':'서대문구 신촌동',
       '11410615':'서대문구 연희동', '11410620':'서대문구 홍제제1동', '11410640':'서대문구 홍제제3동', '11410655':'서대문구 홍제제2동',
       '11410660':'서대문구 홍은제1동', '11410685':'서대문구 홍은제2동', '11410690':'서대문구 남가좌제1동', '11410700':'서대문구 남가좌제2동',
       '11410710':'서대문구 북가좌제1동', '11410720':'서대문구 북가좌제2동',
      
       '11440555':'마포구 아현동', '11440565':'마포구 공덕동', '11440585':'마포구 도화동', '11440590':'마포구 용강동', 
       '11440600':'마포구 대흥동', '11440610':'마포구 염리동', '11440630':'마포구 신수동', '11440655':'마포구 서강동',
       '11440660':'마포구 서교동', '11440680':'마포구 합정동', '11440690':'마포구 망원제1동', '11440700':'마포구 망원제2동',
       '11440710':'마포구 연남동', '11440720':'마포구 성산제1동', '11440730':'마포구 성산제2동', '11440740':'마포구 상암동',
      
       '11470510':'양천구 목1동', '11470520':'양천구 목2동', '11470530':'양천구 목3동', '11470540':'양천구 목4동',
       '11470550':'양천구 목5동', '11470560':'양천구 신월1동', '11470570':'양천구 신월2동', '11470580':'양천구 신월3동',
       '11470590':'양천구 신월4동', '11470600':'양천구 신월5동', '11470610':'양천구 신월6동', '11470611':'양천구 신월7동',
       '11470620':'양천구 신정1동', '11470630':'양천구 신정2동', '11470640':'양천구 신정3동', '11470650':'양천구 신정4동',
       '11470670':'양천구 신정6동', '11470680':'양천구 신정7동',
      
       '11500510':'강서구 염창동', '11500520':'강서구 등촌제1동', '11500530':'강서구 등촌제2동', '11500535':'강서구 등촌제3동',
       '11500540':'강서구 화곡제1동', '11500550':'강서구 화곡제2동', '11500560':'강서구 화곡제3동', '11500570':'강서구 화곡제4동',
       '11500590':'강서구 화곡본동', '11500591':'강서구 화곡제6동', '11500593':'강서구 화곡제8동', '11500603':'강서구 가양제1동',
       '11500604':'강서구 가양제2동', '11500605':'강서구 가양제3동', '11500611':'강서구 발산제1동', '11500615':'강서구 우장산동',
       '11500620':'강서구 공항동', '11500630':'강서구 방화제1동', '11500640':'강서구 방화제2동', '11500641':'강서구 방화제3동',
      
       '11530510':'구로구 신도림동', '11530520':'구로구 구로제1동', '11530530':'구로구 구로제2동', '11530540':'구로구 구로제3동',
       '11530550':'구로구 구로제4동', '11530560':'구로구 구로제5동', '11530595':'구로구 가리봉동', '11530720':'구로구 고척제1동',
       '11530730':'구로구 고척제2동', '11530740':'구로구 개봉제1동', '11530750':'구로구 개봉제2동', '11530760':'구로구 개봉제3동',
       '11530770':'구로구 오류제1동', '11530780':'구로구 오류제2동', '11530790':'구로구 수궁동',
      
       '11545510':'금천구 가산동', '11545610':'금천구 독산제1동', '11545620':'금천구 독산제2동', '11545630':'금천구 독산제3동',
       '11545640':'금천구 독산제4동', '11545670':'금천구 시흥제1동', '11545680':'금천구 시흥제2동', '11545690':'금천구 시흥제3동',
       '11545700':'금천구 시흥제4동', '11545710':'금천구 시흥제5동',
      
       '11560515':'영등포구 영등포본동', '11560535':'영등포구 영등포동', '11560540':'영등포구 여의동', '11560550':'영등포구 당산제1동',
       '11560560':'영등포구 당산제2동', '11560585':'영등포구 도림동', '11560605':'영등포구 문래동', '11560610':'영등포구 양평제1동',
       '11560620':'영등포구 양평제2동', '11560630':'영등포구 신길제1동', '11560650':'영등포구 신길제3동', '11560660':'영등포구 신길제4동',
       '11560670':'영등포구 신길제5동', '11560680':'영등포구 신길제6동', '11560690':'영등포구 신길제7동', '11560700':'영등포구 대림제1동',
       '11560710':'영등포구 대림제2동', '11560720':'영등포구 대림제3동',
      
       '11590510':'동작구 노량진제1동', '11590520':'동작구 노량진제2동', '11590530':'동작구 상도제1동', '11590540':'동작구 상도제2동',
       '11590550':'동작구 상도제3동', '11590560':'동작구 상도제4동', '11590605':'동작구 흑석동', '11590620':'동작구 사당제1동',
       '11590630':'동작구 사당제2동', '11590640':'동작구 사당제3동', '11590650':'동작구 사당제4동', '11590651':'동작구 사당제5동',
       '11590660':'동작구 대방동', '11590670':'동작구 신대방제1동', '11590680':'동작구 신대방제2동',
      
       '11620525':'관악구 보라매동', '11620545':'관악구 청림동', '11620565':'관악구 성현동', '11620575':'관악구 행운동',
       '11620585':'관악구 낙성대동', '11620595':'관악구 청룡동', '11620605':'관악구 은천동', '11620615':'관악구 중앙동',
       '11620625':'관악구 인헌동', '11620630':'관악구 남현동', '11620645':'관악구 서원동', '11620655':'관악구 신원동',
       '11620665':'관악구 서림동', '11620695':'관악구 신림동', '11620715':'관악구 난향동', '11620725':'관악구 조원동', 
       '11620735':'관악구 대학동', '11620745':'관악구 삼성동', '11620765':'관악구 미성동', '11620775':'관악구 난곡동',
      
       '11650510':'서초구 서초1동', '11650520':'서초구 서초2동', '11650530':'서초구 서초3동', '11650531':'서초구 서초4동',
       '11650540':'서초구 잠원동', '11650550':'서초구 반포본동', '11650560':'서초구 반포1동', '11650570':'서초구 반포2동',
       '11650580':'서초구 반포3동', '11650581':'서초구 반포4동', '11650590':'서초구 방배본동', '11650600':'서초구 방배1동',
       '11650610':'서초구 방배2동', '11650620':'서초구 방배3동', '11650621':'서초구 방배4동', '11650651':'서초구 양재1동',
       '11650652':'서초구 양재2동', '11650660':'서초구 내곡동',
      
       '11680510':'강남구 신사동', '11680521':'강남구 논현1동', '11680531':'강남구 논현2동', '11680545':'강남구 압구정동',
       '11680565':'강남구 청담동', '11680580':'강남구 삼성1동', '11680590':'강남구 삼성2동', '11680600':'강남구 대치1동',
       '11680610':'강남구 대치2동', '11680630':'강남구 대치4동', '11680640':'강남구 역삼1동', '11680650':'강남구 역삼2동',
       '11680655':'강남구 도곡1동', '11680656':'강남구 도곡2동', '11680660':'강남구 개포1동', '11680670':'강남구 개포2동',
       '11680690':'강남구 개포4동', '11680700':'강남구 세곡동', '11680720':'강남구 일원본동', '11680730':'강남구 일원1동',
       '11680740':'강남구 일원2동', '11680750':'강남구 수서동',
      
       '11710510':'송파구 풍납1동', '11710520':'송파구 풍납2동', '11710531':'송파구 거여1동', '11710532':'송파구 거여2동',
       '11710540':'송파구 마천1동', '11710550':'송파구 마천2동', '11710561':'송파구 방이1동', '11710562':'송파구 방이2동',
       '11710566':'송파구 오륜동', '11710570':'송파구 오금동', '11710580':'송파구 송파1동', '11710590':'송파구 송파2동',
       '11710600':'송파구 석촌동', '11710610':'송파구 삼전동', '11710620':'송파구 가락본동', '11710631':'송파구 가락1동',
       '11710632':'송파구 가락2동', '11710641':'송파구 문정1동', '11710642':'송파구 문정2동', '11710646':'송파구 장지동',
       '11710647':'송파구 위례동', '11710650':'송파구 잠실본동', '11710670':'송파구 잠실2동', '11710680':'송파구 잠실3동', 
       '11710690':'송파구 잠실4동', '11710710':'송파구 잠실6동', '11710720':'송파구 잠실7동',
      
       '11740515':'강동구 강일동', '11740520':'강동구 상일동', '11740530':'강동구 명일제1동', '11740540':'강동구 명일제2동',
       '11740550':'강동구 고덕제1동', '11740560':'강동구 고덕제2동', '11740570':'강동구 암사제1동', '11740580':'강동구 암사제2동',
       '11740590':'강동구 암사제3동', '11740600':'강동구 천호제1동', '11740610':'강동구 천호제2동', '11740620':'강동구 천호제3동',
       '11740640':'강동구 성내제1동', '11740650':'강동구 성내제2동', '11740660':'강동구 성내제3동', '11740685':'강동구 길동',
       '11740690':'강동구 둔촌제1동', '11740700':'강동구 둔촌제2동'}

In [12]:
dic_df = pd.DataFrame.from_dict(dic, orient = 'index')
dic_df = dic_df.reset_index()
dic_df.rename(columns = {'index': 'dong_code', 0: 'dong'}, inplace = True)
dic_df

dong_code       dong
0    11110515  종로구 청운효자동
1    11110530    종로구 사직동
2    11110540    종로구 삼청동
3    11110550    종로구 부암동
4    11110560    종로구 평창동
..        ...        ...
418  11740650  강동구 성내제2동
419  11740660  강동구 성내제3동
420  11740685     강동구 길동
421  11740690  강동구 둔촌제1동
422  11740700  강동구 둔촌제2동

[423 rows x 2 columns]

- dong 변수를 "종로구_청운효자동" 이런 식으로 변경해주겠다.

In [13]:
result = []
for i in range(len(dic_df)):
    dong_split = dic_df['dong'][i].split()
    result.append('_'.join(dong_split))
    
result

['종로구_청운효자동',
 '종로구_사직동',
 '종로구_삼청동',
 '종로구_부암동',
 '종로구_평창동',
 '종로구_무악동',
 '종로구_교남동',
 '종로구_가회동',
 '종로구_종로1.2.3.4가동',
 '종로구_종로5.6가동',
 '종로구_이화동',
 '종로구_혜화동',
 '종로구_창신제1동',
 '종로구_창신제2동',
 '종로구_창신제3동',
 '종로구_숭인제1동',
 '종로구_숭인제2동',
 '중구_소공동',
 '중구_회현동',
 '중구_명동',
 '중구_필동',
 '중구_장충동',
 '중구_광희동',
 '중구_을지로동',
 '중구_신당동',
 '중구_다산동',
 '중구_약수동',
 '중구_청구동',
 '중구_신당제5동',
 '중구_동화동',
 '중구_황학동',
 '중구_중림동',
 '용산구_후암동',
 '용산구_용산2가동',
 '용산구_남영동',
 '용산구_청파동',
 '용산구_원효로제1동',
 '용산구_원효로제2동',
 '용산구_효창동',
 '용산구_용문동',
 '용산구_한강로동',
 '용산구_이촌제1동',
 '용산구_이촌제2동',
 '용산구_이태원제1동',
 '용산구_이태원제2동',
 '용산구_한남동',
 '용산구_서빙고동',
 '용산구_보광동',
 '성동구_왕십리제2동',
 '성동구_왕십리도선동',
 '성동구_마장동',
 '성동구_사근동',
 '성동구_행당제1동',
 '성동구_행당제2동',
 '성동구_응봉동',
 '성동구_금호1가동',
 '성동구_금호2.3가동',
 '성동구_금호4가동',
 '성동구_옥수동',
 '성동구_성수1가제1동',
 '성동구_성수1가제2동',
 '성동구_성수2가제1동',
 '성동구_성수2가제3동',
 '성동구_송정동',
 '성동구_용답동',
 '광진구_화양동',
 '광진구_군자동',
 '광진구_중곡제1동',
 '광진구_중곡제2동',
 '광진구_중곡제3동',
 '광진구_중곡제4동',
 '광진구_능동',
 '광진구_광장동',
 '광진구_자양제1동',
 '광진구_자양제2동',
 '광진구_자양제3동',
 '광진구_자양제4동

In [14]:
dic_df['dong'] = result
dic_df

dong_code       dong
0    11110515  종로구_청운효자동
1    11110530    종로구_사직동
2    11110540    종로구_삼청동
3    11110550    종로구_부암동
4    11110560    종로구_평창동
..        ...        ...
418  11740650  강동구_성내제2동
419  11740660  강동구_성내제3동
420  11740685     강동구_길동
421  11740690  강동구_둔촌제1동
422  11740700  강동구_둔촌제2동

[423 rows x 2 columns]

- 인구 데이터 셋(pop_total)과 수정된 행정동코드 데이터 셋(dic_df)의 'dong_code' 변수 타입이 서로 다르다.
- 때문에 dic_df 데이터 셋의 'dong_code' 변수를 int형으로 변경해주겠다.

In [15]:
pop_total.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1852032 entries, 0 to 1852031
Data columns (total 11 columns):
 #   Column        Dtype         
---  ------        -----         
 0   date          datetime64[ns]
 1   hour          int64         
 2   dong_code     int64         
 3   total         float64       
 4   pop_below_20  float64       
 5   pop_20        float64       
 6   pop_30        float64       
 7   pop_40        float64       
 8   pop_50        float64       
 9   pop_60        float64       
 10  pop_over_70   float64       
dtypes: datetime64[ns](1), float64(8), int64(2)
memory usage: 155.4 MB


In [16]:
dic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 423 entries, 0 to 422
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   dong_code  423 non-null    object
 1   dong       423 non-null    object
dtypes: object(2)
memory usage: 6.7+ KB


In [17]:
dic_df['dong_code'] = pd.to_numeric(dic_df['dong_code'])
dic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 423 entries, 0 to 422
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   dong_code  423 non-null    int64 
 1   dong       423 non-null    object
dtypes: int64(1), object(1)
memory usage: 6.7+ KB


- pop_total 데이터 셋에는 'dong_code'가 424개가 존재하는데, dic_df 데이터 셋에는 'dong_code'가 423개 존재한다.
    - 최종 inner join한 데이터 셋을 통해, 1개 차이로 인해 len이 불일치함을 확인하였다.
- 따라서 이 부분을 해결해주고 inner join을 수행해주도록 하겠다.
    - 해결 방안은 직접 비교해서 dic_df 데이터 셋에 미존재하는 **"11710647(00)"** 행정동을 직접 찾아서 넣어주어야 한다.
    - 구글링을 해 본 결과, **"11710647(00)"**에 해당하는 행정동은 **"위례동"**임을 확인했다.
    - 그래서 dic(행정동코드 변환 딕셔너리)에 **```'11710647':'위례동'```** 코드를 넣어주었다.
    
**<참고>**
- 위 사항들을 확인해보기 위한 코드는 모두 주석 처리 해두었다.

In [18]:
# pop_total_final = pd.merge(pop_total, dic_df, how = 'inner', on = 'dong_code')
# pop_total_final

In [19]:
# df1 = pd.DataFrame(set(pop_total['dong_code']))
# df2 = pd.DataFrame(set(dic_df['dong_code']))

# df1.rename(columns = {0:'dong_code'}, inplace = True)
# df2.rename(columns = {0:'dong_code'}, inplace = True)

# check1 = df1.sort_values(by = 'dong_code')
# check2 = df2.sort_values(by = 'dong_code')

In [20]:
# # c_total 데이터 셋을 50개의 row 씩 잘라서 눈으로 확인
# # 그 결과, dic_df 데이터 셋에 "11710647" 행정동코드가 미존재함을 확인!!
# c_total = pd.concat([check1, check2], axis = 1)

In [21]:
# pop_total[pop_total['dong_code'] == 11710647]

- 이제 pop_total 데이터 셋과 dic_df 데이터 셋을 'dong_code' 변수를 조인키로 사용하여 inner join 해주겠다.

In [22]:
pop_total_final = pd.merge(pop_total, dic_df, how = 'inner', on = 'dong_code')
pop_total_final

date  hour  dong_code       total  pop_below_20     pop_20  \
0       2020-01-01     0   11620625  24188.5806     3347.7743  4405.0736   
1       2020-01-01     1   11620625  24573.9734     3332.9894  4546.4644   
2       2020-01-01     2   11620625  24817.5792     3294.7822  4705.3521   
3       2020-01-01     3   11620625  24949.4976     3295.4704  4765.7607   
4       2020-01-01     4   11620625  25194.9277     3351.1315  4851.4705   
...            ...   ...        ...         ...           ...        ...   
1847659 2020-06-30    19   11230720  17404.1470     2183.8231  4710.2625   
1847660 2020-06-30    20   11230720  17299.4610     2158.8373  4686.7613   
1847661 2020-06-30    21   11230720  17859.0522     2217.5795  4969.8965   
1847662 2020-06-30    22   11230720  17865.4329     2322.5385  5031.6162   
1847663 2020-06-30    23   11230720  15952.3005     2102.2306  4493.2810   

            pop_30     pop_40     pop_50     pop_60  pop_over_70        dong  
0        4848.7403  3635.8273  3035.2176  2479.2207    2436.7258     관악구_인헌동  
1        4921.4905  3683.1960  3117.8767  2530.5035    2441.4543     관악구_인헌동  
2        4992.3661  3695.5277  3121.4930  2539.1656    2468.8946     관악구_인헌동  
3        5066.2526  3718.6047  3109.1654  2546.5768    2447.6669     관악구_인헌동  
4        5099.5470  3766.5076  3112.5671  2544.5949    2469.1107     관악구_인헌동  
...            ...        ...        ...        ...          ...         ...  
1847659  2355.6257  2151.9566  2343.6567  1762.8175    1896.0046  동대문구_휘경제1동  
1847660  2422.3996  2168.1554  2322.8088  1733.8330    1806.6659  동대문구_휘경제1동  
1847661  2448.4338  2290.0782  2345.2174  1784.8384    1803.0098  동대문구_휘경제1동  
1847662  2420.9131  2234.2893  2337.9373  1741.0912    1777.0481  동대문구_휘경제1동  
1847663  2291.4606  1932.5428  1938.1039  1528.2801    1666.4016  동대문구_휘경제1동  

[1847664 rows x 12 columns]

In [23]:
pop_total_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1847664 entries, 0 to 1847663
Data columns (total 12 columns):
 #   Column        Dtype         
---  ------        -----         
 0   date          datetime64[ns]
 1   hour          int64         
 2   dong_code     int64         
 3   total         float64       
 4   pop_below_20  float64       
 5   pop_20        float64       
 6   pop_30        float64       
 7   pop_40        float64       
 8   pop_50        float64       
 9   pop_60        float64       
 10  pop_over_70   float64       
 11  dong          object        
dtypes: datetime64[ns](1), float64(8), int64(2), object(1)
memory usage: 183.3+ MB


# 인구수 관련 변수들 데이터 타입 변경
- 소수점으로 되어 있는 인구수 관련 변수들을 모두 '올림' 처리하여 정수로 만들어주겠다.

In [24]:
var_list = ['total', 'pop_below_20', 'pop_20', 'pop_30', 'pop_40', 'pop_50', 'pop_60', 'pop_over_70']
pop_total_final[var_list] = np.ceil(pop_total_final[var_list])

In [25]:
pop_total_final.head()

date  hour  dong_code    total  pop_below_20  pop_20  pop_30  pop_40  \
0 2020-01-01     0   11620625  24189.0        3348.0  4406.0  4849.0  3636.0   
1 2020-01-01     1   11620625  24574.0        3333.0  4547.0  4922.0  3684.0   
2 2020-01-01     2   11620625  24818.0        3295.0  4706.0  4993.0  3696.0   
3 2020-01-01     3   11620625  24950.0        3296.0  4766.0  5067.0  3719.0   
4 2020-01-01     4   11620625  25195.0        3352.0  4852.0  5100.0  3767.0   

   pop_50  pop_60  pop_over_70     dong  
0  3036.0  2480.0       2437.0  관악구_인헌동  
1  3118.0  2531.0       2442.0  관악구_인헌동  
2  3122.0  2540.0       2469.0  관악구_인헌동  
3  3110.0  2547.0       2448.0  관악구_인헌동  
4  3113.0  2545.0       2470.0  관악구_인헌동

# 조인키 생성
- 지하철 데이터 셋을 기준으로 인구 데이터 셋과 기상 데이터 셋을 결합시켜 줄 것이므로, 이를 위한 조인키가 필요하다.
- 인구 데이터의 'dong' 변수의 경우, "종로구_청운효자동" 이런 식으로 자치구와 행정동이 같이 입력되어 있다.
- 따라서 다음과 같이 조인키를 생성해주겠다.
    1. 앞서 했던 방법대로 먼저 문자열을 결합해준다.
        - ex) "2018-01-01_종로구_청운효자동"
    2. 만들어진 조인키 변수에서 자치구에 해당 되는 부분은 제거해준다.
        - ex) "2018-01-01_청운효자동"

In [26]:
# datetime to string
pop_total_final['date'] = pop_total_final['date'].astype(str)

# 문자열 결합
pop_total_final['pop_join_key'] = pop_total_final['date'] + '_' + pop_total_final['dong']
# '_'를 기준으로 문자열을 잘라서 리스트에 담기
pop_join_key_list = pop_total_final['pop_join_key'].str.split('_').to_list()

In [27]:
# '자치구' 부분을 제외해 준 뒤에 문자열 재결합
result = []
for i in tqdm(pop_join_key_list):
    result.append('_'.join([i[0], i[2]]))

# 지하철 데이터와 결합할 때 사용할 조인키
pop_total_final['pop_join_key'] = result
pop_total_final['pop_join_key']

0            2020-01-01_인헌동
1            2020-01-01_인헌동
2            2020-01-01_인헌동
3            2020-01-01_인헌동
4            2020-01-01_인헌동
                 ...       
1847659    2020-06-30_휘경제1동
1847660    2020-06-30_휘경제1동
1847661    2020-06-30_휘경제1동
1847662    2020-06-30_휘경제1동
1847663    2020-06-30_휘경제1동
Name: pop_join_key, Length: 1847664, dtype: object

# 불필요한 변수들 제거
- dong_code, dong 변수는 분석에 사용하지 않을 것이므로 제거해주겠다.

In [28]:
pop_total_final.columns

Index(['date', 'hour', 'dong_code', 'total', 'pop_below_20', 'pop_20',
       'pop_30', 'pop_40', 'pop_50', 'pop_60', 'pop_over_70', 'dong',
       'pop_join_key'],
      dtype='object')

In [29]:
pop_total_final.drop(['dong_code', 'dong'], axis = 1, inplace = True)

In [30]:
pop_total_final

date  hour    total  pop_below_20  pop_20  pop_30  pop_40  \
0        2020-01-01     0  24189.0        3348.0  4406.0  4849.0  3636.0   
1        2020-01-01     1  24574.0        3333.0  4547.0  4922.0  3684.0   
2        2020-01-01     2  24818.0        3295.0  4706.0  4993.0  3696.0   
3        2020-01-01     3  24950.0        3296.0  4766.0  5067.0  3719.0   
4        2020-01-01     4  25195.0        3352.0  4852.0  5100.0  3767.0   
...             ...   ...      ...           ...     ...     ...     ...   
1847659  2020-06-30    19  17405.0        2184.0  4711.0  2356.0  2152.0   
1847660  2020-06-30    20  17300.0        2159.0  4687.0  2423.0  2169.0   
1847661  2020-06-30    21  17860.0        2218.0  4970.0  2449.0  2291.0   
1847662  2020-06-30    22  17866.0        2323.0  5032.0  2421.0  2235.0   
1847663  2020-06-30    23  15953.0        2103.0  4494.0  2292.0  1933.0   

         pop_50  pop_60  pop_over_70      pop_join_key  
0        3036.0  2480.0       2437.0    2020-01-01_인헌동  
1        3118.0  2531.0       2442.0    2020-01-01_인헌동  
2        3122.0  2540.0       2469.0    2020-01-01_인헌동  
3        3110.0  2547.0       2448.0    2020-01-01_인헌동  
4        3113.0  2545.0       2470.0    2020-01-01_인헌동  
...         ...     ...          ...               ...  
1847659  2344.0  1763.0       1897.0  2020-06-30_휘경제1동  
1847660  2323.0  1734.0       1807.0  2020-06-30_휘경제1동  
1847661  2346.0  1785.0       1804.0  2020-06-30_휘경제1동  
1847662  2338.0  1742.0       1778.0  2020-06-30_휘경제1동  
1847663  1939.0  1529.0       1667.0  2020-06-30_휘경제1동  

[1847664 rows x 11 columns]

In [31]:
pop_total_final.isnull().sum()

date            0
hour            0
total           0
pop_below_20    0
pop_20          0
pop_30          0
pop_40          0
pop_50          0
pop_60          0
pop_over_70     0
pop_join_key    0
dtype: int64

In [32]:
pop_total_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1847664 entries, 0 to 1847663
Data columns (total 11 columns):
 #   Column        Dtype  
---  ------        -----  
 0   date          object 
 1   hour          int64  
 2   total         float64
 3   pop_below_20  float64
 4   pop_20        float64
 5   pop_30        float64
 6   pop_40        float64
 7   pop_50        float64
 8   pop_60        float64
 9   pop_over_70   float64
 10  pop_join_key  object 
dtypes: float64(8), int64(1), object(2)
memory usage: 169.2+ MB


# 최종 생성된 데이터 셋을 csv 파일로 내보내기

In [33]:
new_filepath = 'C:/Users/Brian/Desktop/Data Science/Projects/Playdata/dataset/made_by_전처리/test'

In [34]:
pop_total_final.to_csv('{}/pop_final.csv'.format(new_filepath), index = False)